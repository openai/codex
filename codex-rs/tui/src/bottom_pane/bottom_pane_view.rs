use crate::user_approval_widget::ApprovalRequest;
use crossterm::event::KeyEvent;
use ratatui::buffer::Buffer;
use ratatui::layout::Rect;

use super::BottomPane;
use super::CancellationEvent;

/// Trait implemented by every view that can be shown in the bottom pane.
pub(crate) trait BottomPaneView {
    /// Handle a key event while the view is active. A redraw is always
    /// scheduled after this call.
    fn handle_key_event(&mut self, _pane: &mut BottomPane, _key_event: KeyEvent) {}

    /// Return `true` if the view has finished and should be removed.
    fn is_complete(&self) -> bool {
        false
    }

    /// Handle Ctrl-C while this view is active.
    fn on_ctrl_c(&mut self, _pane: &mut BottomPane) -> CancellationEvent {
        CancellationEvent::NotHandled
    }

    /// Return the desired height of the view.
    fn desired_height(&self, width: u16) -> u16;

    /// Render the view: this will be displayed in place of the composer.
    fn render(&self, area: Rect, buf: &mut Buffer);

    /// Optional paste handler. Return true if the view modified its state and
    /// needs a redraw.
    fn handle_paste(&mut self, _pane: &mut BottomPane, _pasted: String) -> bool {
        false
    }

    /// Cursor position when this view is active.
    fn cursor_pos(&self, _area: Rect) -> Option<(u16, u16)> {
        None
    }

    /// Try to handle approval request; return the original value if not
    /// consumed.
    fn try_consume_approval_request(
        &mut self,
        request: ApprovalRequest,
    ) -> Option<ApprovalRequest> {
        Some(request)
    }
}
