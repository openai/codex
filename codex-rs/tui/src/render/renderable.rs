use std::sync::Arc;

use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use ratatui::text::Line;
use ratatui::text::Span;
use ratatui::widgets::Paragraph;
use ratatui::widgets::WidgetRef;

use crate::render::Insets;
use crate::render::RectExt as _;

pub trait Renderable {
    fn render(&self, area: Rect, buf: &mut Buffer);
    fn desired_height(&self, width: u16) -> u16;
    fn cursor_pos(&self, _area: Rect) -> Option<(u16, u16)> {
        None
    }
}

pub enum RenderableItem<'a> {
    Owned(Box<dyn Renderable + 'a>),
    Borrowed(&'a dyn Renderable),
}

impl<'a> Renderable for RenderableItem<'a> {
    fn render(&self, area: Rect, buf: &mut Buffer) {
        match self {
            RenderableItem::Owned(child) => child.render(area, buf),
            RenderableItem::Borrowed(child) => child.render(area, buf),
        }
    }

    fn desired_height(&self, width: u16) -> u16 {
        match self {
            RenderableItem::Owned(child) => child.desired_height(width),
            RenderableItem::Borrowed(child) => child.desired_height(width),
        }
    }

    fn cursor_pos(&self, area: Rect) -> Option<(u16, u16)> {
        match self {
            RenderableItem::Owned(child) => child.cursor_pos(area),
            RenderableItem::Borrowed(child) => child.cursor_pos(area),
        }
    }
}

impl<'a> From<Box<dyn Renderable + 'a>> for RenderableItem<'a> {
    fn from(value: Box<dyn Renderable + 'a>) -> Self {
        RenderableItem::Owned(value)
    }
}

impl<'a, R> From<R> for Box<dyn Renderable + 'a>
where
    R: Renderable + 'a,
{
    fn from(value: R) -> Self {
        Box::new(value)
    }
}

impl Renderable for () {
    fn render(&self, _area: Rect, _buf: &mut Buffer) {}
    fn desired_height(&self, _width: u16) -> u16 {
        0
    }
}

impl Renderable for &str {
    fn render(&self, area: Rect, buf: &mut Buffer) {
        self.render_ref(area, buf);
    }
    fn desired_height(&self, _width: u16) -> u16 {
        1
    }
}

impl Renderable for String {
    fn render(&self, area: Rect, buf: &mut Buffer) {
        self.render_ref(area, buf);
    }
    fn desired_height(&self, _width: u16) -> u16 {
        1
    }
}

impl<'a> Renderable for Span<'a> {
    fn render(&self, area: Rect, buf: &mut Buffer) {
        self.render_ref(area, buf);
    }
    fn desired_height(&self, _width: u16) -> u16 {
        1
    }
}

impl<'a> Renderable for Line<'a> {
    fn render(&self, area: Rect, buf: &mut Buffer) {
        WidgetRef::render_ref(self, area, buf);
    }
    fn desired_height(&self, _width: u16) -> u16 {
        1
    }
}

impl<'a> Renderable for Paragraph<'a> {
    fn render(&self, area: Rect, buf: &mut Buffer) {
        self.render_ref(area, buf);
    }
    fn desired_height(&self, width: u16) -> u16 {
        self.line_count(width) as u16
    }
}

impl<R: Renderable> Renderable for Option<R> {
    fn render(&self, area: Rect, buf: &mut Buffer) {
        if let Some(renderable) = self {
            renderable.render(area, buf);
        }
    }

    fn desired_height(&self, width: u16) -> u16 {
        if let Some(renderable) = self {
            renderable.desired_height(width)
        } else {
            0
        }
    }
}

impl<R: Renderable> Renderable for Arc<R> {
    fn render(&self, area: Rect, buf: &mut Buffer) {
        self.as_ref().render(area, buf);
    }
    fn desired_height(&self, width: u16) -> u16 {
        self.as_ref().desired_height(width)
    }
}

pub struct ColumnRenderable<'a> {
    children: Vec<RenderableItem<'a>>,
}

impl Renderable for ColumnRenderable<'_> {
    fn render(&self, area: Rect, buf: &mut Buffer) {
        let mut y = area.y;
        for child in &self.children {
            let child_area = Rect::new(area.x, y, area.width, child.desired_height(area.width))
                .intersection(area);
            if !child_area.is_empty() {
                child.render(child_area, buf);
            }
            y += child_area.height;
        }
    }

    fn desired_height(&self, width: u16) -> u16 {
        self.children
            .iter()
            .map(|child| child.desired_height(width))
            .sum()
    }

    fn cursor_pos(&self, area: Rect) -> Option<(u16, u16)> {
        let mut y = area.y;
        for child in &self.children {
            let child_area = Rect::new(area.x, y, area.width, child.desired_height(area.width))
                .intersection(area);
            if !child_area.is_empty()
                && let Some((px, py)) = child.cursor_pos(child_area)
            {
                return Some((px, py));
            }
            y += child_area.height;
        }
        None
    }
}

impl<'a> ColumnRenderable<'a> {
    pub fn new() -> Self {
        Self { children: vec![] }
    }

    pub fn with<I, T>(children: I) -> Self
    where
        I: IntoIterator<Item = T>,
        T: Into<RenderableItem<'a>>,
    {
        Self {
            children: children.into_iter().map(Into::into).collect(),
        }
    }

    pub fn push(&mut self, child: impl Into<Box<dyn Renderable + 'a>>) {
        self.children.push(RenderableItem::Owned(child.into()));
    }

    #[allow(dead_code)]
    pub fn push_ref<R>(&mut self, child: &'a R)
    where
        R: Renderable + 'a,
    {
        self.children
            .push(RenderableItem::Borrowed(child as &'a dyn Renderable));
    }
}

pub struct FlexChild<'a> {
    flex: i32,
    child: RenderableItem<'a>,
}

pub struct FlexRenderable<'a> {
    children: Vec<FlexChild<'a>>,
}

impl<'a> FlexRenderable<'a> {
    pub fn new() -> Self {
        Self { children: vec![] }
    }

    pub fn push(&mut self, flex: i32, child: impl Into<RenderableItem<'a>>) {
        self.children.push(FlexChild {
            flex,
            child: child.into(),
        });
    }

    fn allocate(&self, area: Rect) -> Vec<Rect> {
        let mut allocated_rects = Vec::with_capacity(self.children.len());
        let mut child_sizes = vec![0; self.children.len()];
        let mut allocated_size = 0;
        let mut total_flex = 0;

        let max_size = area.height;
        let mut last_flex_child_idx = 0;
        for (i, FlexChild { flex, child }) in self.children.iter().enumerate() {
            if *flex > 0 {
                total_flex += flex;
                last_flex_child_idx = i;
            } else {
                child_sizes[i] = child
                    .desired_height(area.width)
                    .min(max_size.saturating_sub(allocated_size));
                allocated_size += child_sizes[i];
            }
        }
        let free_space = max_size.saturating_sub(allocated_size);
        let mut allocated_flex_space = 0;
        if total_flex > 0 {
            let space_per_flex = free_space / total_flex as u16;
            for (i, FlexChild { flex, child }) in self.children.iter().enumerate() {
                if *flex > 0 {
                    let max_child_extent = if i == last_flex_child_idx {
                        free_space - allocated_flex_space
                    } else {
                        space_per_flex * *flex as u16
                    };
                    let child_size = child.desired_height(area.width).min(max_child_extent);
                    child_sizes[i] = child_size;
                    allocated_size += child_size;
                    allocated_flex_space += child_size;
                }
            }
        }

        let mut y = area.y;
        for size in child_sizes {
            let child_area = Rect::new(area.x, y, area.width, size);
            allocated_rects.push(child_area);
            y += child_area.height;
        }
        allocated_rects
    }
}

impl<'a> Renderable for FlexRenderable<'a> {
    fn render(&self, area: Rect, buf: &mut Buffer) {
        self.allocate(area)
            .into_iter()
            .zip(self.children.iter())
            .for_each(|(rect, child)| {
                child.child.render(rect, buf);
            });
    }

    fn desired_height(&self, width: u16) -> u16 {
        self.allocate(Rect::new(0, 0, width, u16::MAX))
            .last()
            .map(|rect| rect.bottom())
            .unwrap_or(0)
    }

    fn cursor_pos(&self, area: Rect) -> Option<(u16, u16)> {
        self.allocate(area)
            .into_iter()
            .zip(self.children.iter())
            .find_map(|(rect, child)| child.child.cursor_pos(rect))
    }
}

pub struct RowRenderable<'a> {
    children: Vec<(u16, RenderableItem<'a>)>,
}

impl Renderable for RowRenderable<'_> {
    fn render(&self, area: Rect, buf: &mut Buffer) {
        let mut x = area.x;
        for (width, child) in &self.children {
            let available_width = area.width.saturating_sub(x - area.x);
            let child_area = Rect::new(x, area.y, (*width).min(available_width), area.height);
            if child_area.is_empty() {
                break;
            }
            child.render(child_area, buf);
            x = x.saturating_add(*width);
        }
    }
    fn desired_height(&self, width: u16) -> u16 {
        let mut max_height = 0;
        let mut width_remaining = width;
        for (child_width, child) in &self.children {
            let w = (*child_width).min(width_remaining);
            if w == 0 {
                break;
            }
            let height = child.desired_height(w);
            if height > max_height {
                max_height = height;
            }
            width_remaining = width_remaining.saturating_sub(w);
        }
        max_height
    }

    fn cursor_pos(&self, area: Rect) -> Option<(u16, u16)> {
        let mut x = area.x;
        for (width, child) in &self.children {
            let available_width = area.width.saturating_sub(x - area.x);
            let child_area = Rect::new(x, area.y, (*width).min(available_width), area.height);
            if !child_area.is_empty()
                && let Some(pos) = child.cursor_pos(child_area)
            {
                return Some(pos);
            }
            x = x.saturating_add(*width);
        }
        None
    }
}

impl<'a> RowRenderable<'a> {
    pub fn new() -> Self {
        Self { children: vec![] }
    }

    pub fn push(&mut self, width: u16, child: impl Into<Box<dyn Renderable>>) {
        self.children
            .push((width, RenderableItem::Owned(child.into())));
    }

    #[allow(dead_code)]
    pub fn push_ref<R>(&mut self, width: u16, child: &'a R)
    where
        R: Renderable + 'a,
    {
        self.children
            .push((width, RenderableItem::Borrowed(child as &'a dyn Renderable)));
    }
}

pub struct InsetRenderable<'a> {
    child: RenderableItem<'a>,
    insets: Insets,
}

impl<'a> Renderable for InsetRenderable<'a> {
    fn render(&self, area: Rect, buf: &mut Buffer) {
        self.child.render(area.inset(self.insets), buf);
    }
    fn desired_height(&self, width: u16) -> u16 {
        self.child
            .desired_height(width - self.insets.left - self.insets.right)
            + self.insets.top
            + self.insets.bottom
    }
    fn cursor_pos(&self, area: Rect) -> Option<(u16, u16)> {
        self.child.cursor_pos(area.inset(self.insets))
    }
}

impl<'a> InsetRenderable<'a> {
    pub fn new(child: impl Into<RenderableItem<'a>>, insets: Insets) -> Self {
        Self {
            child: child.into(),
            insets,
        }
    }
}

pub trait RenderableExt<'a> {
    fn inset(self, insets: Insets) -> RenderableItem<'a>;
}

impl<'a, R> RenderableExt<'a> for R
where
    R: Renderable + 'a,
{
    fn inset(self, insets: Insets) -> RenderableItem<'a> {
        let child: RenderableItem<'a> =
            RenderableItem::Owned(Box::new(self) as Box<dyn Renderable + 'a>);
        RenderableItem::Owned(Box::new(InsetRenderable { child, insets }))
    }
}
