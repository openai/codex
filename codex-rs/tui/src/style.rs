use crate::color::blend;
use crate::color::is_light;
use crate::color::perceptual_distance;
use crate::terminal_palette::basic_palette;
use crate::terminal_palette::terminal_palette;
use ratatui::style::Color;
use ratatui::style::Style;

/// Returns the style for a user-authored message using the provided terminal background.
pub fn user_message_style(terminal_bg: Option<(u8, u8, u8)>) -> Style {
    match terminal_bg {
        Some(bg) => Style::default().bg(user_message_bg(bg)),
        None => Style::default(),
    }
}

#[allow(clippy::disallowed_methods)]
pub fn user_message_bg(terminal_bg: (u8, u8, u8)) -> Color {
    // Determine a reference "top" color to blend toward. For dark backgrounds we lighten
    // the mix with white, and for light backgrounds we darken it with black.
    let top = if is_light(terminal_bg) {
        (0, 0, 0)
    } else {
        (255, 255, 255)
    };
    let bottom = terminal_bg;
    let Some(mut color_level) = supports_color::on_cached(supports_color::Stream::Stdout) else {
        return Color::default();
    };

    #[cfg(windows)]
    // Windows Terminal has been the default shell application for Windows since October 2022
    // and has supported truecolor even longer. However it usually does not set COLORTERM to indicate that.
    // so this is a pretty safe heuristic.
    if std::env::var_os("WT_SESSION").is_some() {
        color_level.has_16m = true;
    }

    let target = blend(top, bottom, 0.1);
    if color_level.has_16m {
        // In truecolor terminals we can use the exact RGB value.
        let (r, g, b) = target;
        Color::Rgb(r, g, b)
    } else if color_level.has_256
        && let Some(palette) = terminal_palette()
        && let Some((i, _)) = palette.into_iter().enumerate().min_by(|(_, a), (_, b)| {
            perceptual_distance(*a, target)
                .partial_cmp(&perceptual_distance(*b, target))
                .unwrap_or(std::cmp::Ordering::Equal)
        })
    {
        // If we have a captured 256-color palette, pick whichever indexed color is
        // perceptually closest to the blended target.
        Color::Indexed(i as u8)
    } else if color_level.has_basic {
        if let Some(palette) = basic_palette() {
            // On Windows terminals the palette is configurable, so evaluate the actual
            // runtime color table to keep the blended shading aligned with custom themes.
            closest_runtime_basic_color(target, terminal_bg, &palette)
        } else {
            // Finally, degrade to the well-known ANSI 16-color defaults using a perceptual
            // distance match.
            closest_basic_color(target, terminal_bg)
        }
    } else {
        Color::default()
    }
}

fn closest_runtime_basic_color(
    target: (u8, u8, u8),
    terminal_bg: (u8, u8, u8),
    palette: &[(u8, u8, u8); 16],
) -> Color {
    select_basic_palette_color(
        target,
        terminal_bg,
        palette.iter().enumerate().filter_map(|(idx, rgb)| {
            BASIC_TERMINAL_COLORS
                .get(idx)
                .map(|(color, _)| (*rgb, *color))
        }),
    )
}

fn closest_basic_color(target: (u8, u8, u8), terminal_bg: (u8, u8, u8)) -> Color {
    // Iterate through the baked-in ANSI colors and return whichever one is closest to the
    // desired RGB shade while maintaining contrast with the background.
    select_basic_palette_color(
        target,
        terminal_bg,
        BASIC_TERMINAL_COLORS
            .iter()
            .map(|(color, rgb)| (*rgb, *color)),
    )
}

const MIN_PERCEPTUAL_DISTANCE: f32 = 6.0;

fn select_basic_palette_color(
    target: (u8, u8, u8),
    terminal_bg: (u8, u8, u8),
    entries: impl Iterator<Item = ((u8, u8, u8), Color)>,
) -> Color {
    let mut best = None;
    let mut fallback = None;
    for (rgb, color) in entries {
        let dist = perceptual_distance(rgb, target);
        if fallback.is_none_or(|(_, best_dist)| dist < best_dist) {
            fallback = Some((color, dist));
        }
        if perceptual_distance(rgb, terminal_bg) > MIN_PERCEPTUAL_DISTANCE
            && best.is_none_or(|(_, best_dist)| dist < best_dist)
        {
            best = Some((color, dist));
        }
    }
    best.or(fallback)
        .map(|(color, _)| color)
        .unwrap_or(Color::default())
}

// Mapping of ANSI color indices to approximate RGB tuples. These values mirror Windows'
// default console palette so the fallback path stays visually consistent across platforms.
const BASIC_TERMINAL_COLORS: [(Color, (u8, u8, u8)); 16] = [
    (Color::Black, (0, 0, 0)),
    (Color::Blue, (0, 0, 128)),
    (Color::Green, (0, 128, 0)),
    (Color::Cyan, (0, 128, 128)),
    (Color::Red, (128, 0, 0)),
    (Color::Magenta, (128, 0, 128)),
    (Color::Yellow, (128, 128, 0)),
    (Color::Gray, (192, 192, 192)),
    (Color::DarkGray, (128, 128, 128)),
    (Color::LightBlue, (0, 0, 255)),
    (Color::LightGreen, (0, 255, 0)),
    (Color::LightCyan, (0, 255, 255)),
    (Color::LightRed, (255, 0, 0)),
    (Color::LightMagenta, (255, 0, 255)),
    (Color::LightYellow, (255, 255, 0)),
    (Color::White, (255, 255, 255)),
];
