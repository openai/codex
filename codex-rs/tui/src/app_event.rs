use std::path::PathBuf;

use codex_common::model_presets::ModelPreset;
use codex_core::protocol::ConversationPathResponseEvent;
use codex_core::protocol::Event;
use codex_file_search::FileMatch;

use crate::bottom_pane::ApprovalRequest;
use crate::history_cell::HistoryCell;

use codex_core::protocol::AskForApproval;
use codex_core::protocol::SandboxPolicy;
use codex_core::protocol_config_types::ReasoningEffort;

#[allow(clippy::large_enum_variant)]
#[derive(Debug)]
pub(crate) enum AppEvent {
    CodexEvent(Event),

    /// Start a new session.
    NewSession,

    /// Request to exit the application gracefully.
    ExitRequest,

    /// Forward an `Op` to the Agent. Using an `AppEvent` for this avoids
    /// bubbling channels through layers of widgets.
    CodexOp(codex_core::protocol::Op),

    /// Kick off an asynchronous file search for the given query (text after
    /// the `@`). Previous searches may be cancelled by the app layer so there
    /// is at most one in-flight search.
    StartFileSearch(String),

    /// Result of a completed asynchronous file search. The `query` echoes the
    /// original search term so the UI can decide whether the results are
    /// still relevant.
    FileSearchResult {
        query: String,
        matches: Vec<FileMatch>,
    },

    /// Result of computing a `/diff` command.
    DiffResult(String),

    InsertHistoryCell(Box<dyn HistoryCell>),

    StartCommitAnimation,
    StopCommitAnimation,
    CommitTick,

    /// Update the current reasoning effort in the running app and widget.
    UpdateReasoningEffort(Option<ReasoningEffort>),

    /// Update the current model slug in the running app and widget.
    UpdateModel(String),

    /// Persist the selected model and reasoning effort to the appropriate config.
    PersistModelSelection {
        model: String,
        effort: Option<ReasoningEffort>,
    },

    /// Open the reasoning selection popup after picking a model.
    OpenReasoningPopup {
        model: String,
        presets: Vec<ModelPreset>,
    },

    /// Update the current approval policy in the running app and widget.
    UpdateAskForApprovalPolicy(AskForApproval),

    /// Update the current sandbox policy in the running app and widget.
    UpdateSandboxPolicy(SandboxPolicy),

    /// Forwarded conversation history snapshot from the current conversation.
    ConversationHistory(ConversationPathResponseEvent),
    /// Resume the conversation from the backup rollout (\*.bak).
    RestoreContextFromBackup,
    /// On shutdown, back up and rewrite the current rollout file according to the
    /// current prune mask so the next resume starts from a pruned file.
    FinalizePruneOnShutdown,

    /// Open the branch picker option from the review popup.
    OpenReviewBranchPicker(PathBuf),

    /// Open the commit picker option from the review popup.
    OpenReviewCommitPicker(PathBuf),

    /// Open the custom prompt option from the review popup.
    OpenReviewCustomPrompt,

    // --- Advanced Prune (experimental) ---
    /// Open the non-destructive advanced prune view.
    OpenPruneAdvanced,
    /// Open the manual prune submenu (by category).
    OpenPruneManual,
    /// Confirm the currently selected manual prune category.
    ConfirmManualChanges,
    /// Apply the pending manual prune plan.
    ApplyManualPrune,
    /// Open the root prune menu.
    OpenPruneRoot,
    /// Open a confirmation dialog for manual prune of a specific category.
    OpenPruneManualConfirm {
        category: codex_core::protocol::PruneCategory,
        label: String,
    },
    /// Close the advanced prune view.
    PruneAdvancedClosed,
    /// Root prune menu dismissed (via Esc or completion).
    PruneRootClosed,
    /// Toggle keep marker for an index in the advanced list.
    ToggleKeepIndex {
        idx: usize,
    },
    /// Toggle delete marker for an index in the advanced list.
    ToggleDeleteIndex {
        idx: usize,
    },
    /// Apply staged inclusion/deletion changes.
    ConfirmAdvancedChanges,
    /// Apply staged advanced prune after user confirmation.
    ApplyAdvancedPrune,

    /// Show an informational toast/message in the transcript.
    ShowInfoMessage(String),

    /// Open the approval popup.
    FullScreenApprovalRequest(ApprovalRequest),

    // (upstream event placeholders removed; handled by our prune events above)
}
