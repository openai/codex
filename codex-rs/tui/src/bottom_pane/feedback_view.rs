use crate::app_event::AppEvent;
use crate::app_event_sender::AppEventSender;
use crate::bottom_pane::bottom_pane_view::BottomPaneView;
use crate::history_cell;
use crate::history_cell::PlainHistoryCell;
use crate::render::Insets;
use crate::render::RectExt as _;
use crate::render::renderable::Renderable;
use crate::style::user_message_style;
use codex_protocol::ConversationId;
use crossterm::event::KeyCode;
use crossterm::event::KeyEvent;
use crossterm::event::KeyEventKind;
use ratatui::buffer::Buffer;
use ratatui::layout::Constraint;
use ratatui::layout::Layout;
use ratatui::layout::Rect;
use ratatui::style::Stylize;
use ratatui::text::Line;
use ratatui::widgets::Block;
use ratatui::widgets::Widget;
use std::path::PathBuf;

pub(crate) struct FeedbackView {
    app_event_tx: AppEventSender,
    file_path: PathBuf,
    snapshot: codex_feedback::CodexLogSnapshot,
    session_id: Option<ConversationId>,
    complete: bool,
}

impl FeedbackView {
    pub fn new(
        app_event_tx: AppEventSender,
        file_path: PathBuf,
        snapshot: codex_feedback::CodexLogSnapshot,
        session_id: Option<ConversationId>,
    ) -> Self {
        Self {
            app_event_tx,
            file_path,
            snapshot,
            session_id,
            complete: false,
        }
    }

    fn header_lines(&self) -> Vec<Line<'static>> {
        vec![
            Line::from("Upload Codex logs to Sentry".bold()),
            "".into(),
            Line::from(
                "Logs might contain the entire conversion history of this Codex process (prompt, tools calls and their results).",
            ),
            Line::from(
                "Logs are persisted for 90 days and are used exclusively to help diagnose issues.",
            ),
            "".into(),
            Line::from(vec![
                "You can inspect the exact content of logs to be uploaded at:".into(),
            ]),
            Line::from(self.file_path.display().to_string().dim()),
            "".into(),
            Line::from("Press Enter to upload to Sentry, or Esc to cancel".dim()),
        ]
    }

    fn render_inner(&self, area: Rect, buf: &mut Buffer) {
        let header_height: u16 = self
            .header_lines()
            .iter()
            .map(|l| l.desired_height(area.width))
            .sum();
        let [header_rect] = Layout::vertical([Constraint::Length(header_height)]).areas(area);

        for (i, line) in self.header_lines().into_iter().enumerate() {
            let line_area = Rect::new(
                header_rect.x,
                header_rect.y + i as u16,
                header_rect.width,
                1,
            )
            .intersection(header_rect);
            line.render(line_area, buf);
        }
    }
}

impl Renderable for FeedbackView {
    fn render(&self, area: Rect, buf: &mut Buffer) {
        if area.width == 0 || area.height == 0 {
            return;
        }
        Block::default()
            .style(user_message_style())
            .render(area, buf);
        let inner = area.inset(Insets::vh(1, 2));
        self.render_inner(inner, buf);
    }

    fn desired_height(&self, width: u16) -> u16 {
        let insets = Insets::vh(1, 2);
        let inner_width = width.saturating_sub(insets.left + insets.right);
        let header_height: u16 = self
            .header_lines()
            .iter()
            .map(|l| l.desired_height(inner_width))
            .sum();
        header_height + insets.top + insets.bottom
    }
}

impl BottomPaneView for FeedbackView {
    fn handle_key_event(&mut self, key_event: KeyEvent) {
        if key_event.kind != KeyEventKind::Press {
            return;
        }
        match key_event.code {
            KeyCode::Enter => {
                match self.snapshot.upload_to_sentry(self.session_id) {
                    Ok(()) => {
                        let issue_url = format!(
                            "https://github.com/openai/codex/issues/new?template=2-bug-report.yml&steps=Uploaded%20thread:%20{}",
                            self.session_id.unwrap_or_default()
                        );

                        self.app_event_tx.send(AppEvent::InsertHistoryCell(Box::new(
                            PlainHistoryCell::new(
                                vec![
                                    Line::from("â€¢ Uploaded Codex logs to Sentry. Please open an issue using the following URL:"), 
                                    "".into(),
                                Line::from(vec!["  ".into(), issue_url.cyan().underlined()])
                                ],
                            )),
                        ));
                    }
                    Err(e) => {
                        self.app_event_tx.send(AppEvent::InsertHistoryCell(Box::new(
                            history_cell::new_error_event(format!(
                                "Failed to upload feedback logs: {e}"
                            )),
                        )));
                    }
                }
                self.complete = true;
            }
            KeyCode::Esc => {
                self.complete = true;
            }
            _ => {}
        }
    }

    fn is_complete(&self) -> bool {
        self.complete
    }

    fn cursor_pos(&self, _area: Rect) -> Option<(u16, u16)> {
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::app_event::AppEvent;
    use crate::style::user_message_style;
    use codex_feedback::CodexFeedback;
    use insta::assert_snapshot;
    use ratatui::Terminal;
    use ratatui::backend::TestBackend;
    use ratatui::layout::Rect;
    use ratatui::style::Color;
    use std::path::PathBuf;
    use tokio::sync::mpsc::unbounded_channel;

    fn buffer_to_string(buffer: &Buffer) -> String {
        (0..buffer.area.height)
            .map(|row| {
                let mut line = String::new();
                for col in 0..buffer.area.width {
                    let symbol = buffer[(buffer.area.x + col, buffer.area.y + row)].symbol();
                    if symbol.is_empty() {
                        line.push(' ');
                    } else {
                        line.push_str(symbol);
                    }
                }
                line.trim_end().to_string()
            })
            .collect::<Vec<_>>()
            .join("\n")
    }

    #[test]
    fn renders_feedback_view_header() {
        let (tx_raw, _rx) = unbounded_channel::<AppEvent>();
        let app_event_tx = AppEventSender::new(tx_raw);
        let snapshot = CodexFeedback::new().snapshot();
        let file_path = PathBuf::from("/tmp/codex-feedback.log");

        let view = FeedbackView::new(
            app_event_tx,
            file_path.clone(),
            snapshot,
            /* session_id */ None,
        );

        let width = 72;
        let height = view.desired_height(width).max(1);
        let mut terminal = Terminal::new(TestBackend::new(width, height)).unwrap();
        terminal
            .draw(|frame| {
                let area = Rect::new(0, 0, width, height);
                view.render(area, frame.buffer_mut());
            })
            .unwrap();

        let rendered = buffer_to_string(terminal.backend().buffer())
            .replace(&file_path.display().to_string(), "<LOG_PATH>");
        assert_snapshot!("feedback_view_render", rendered);

        let area = Rect::new(0, 0, width, height);
        let mut buf = Buffer::empty(area);
        view.render(area, &mut buf);

        let cell_style = buf[(area.x, area.y)].style();
        let expected_bg = user_message_style().bg.unwrap_or(Color::Reset);
        assert_eq!(cell_style.bg.unwrap_or(Color::Reset), expected_bg);
    }
}
