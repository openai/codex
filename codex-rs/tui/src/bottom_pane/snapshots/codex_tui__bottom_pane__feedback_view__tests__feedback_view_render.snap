---
source: tui/src/bottom_pane/feedback_view.rs
assertion_line: 204
expression: normalized
---
  Upload Codex logs to Sentry

  Logs might contain the entire conversion history of this Codex proce
  Logs are persisted for 90 days and are used exclusively to help diag

  You can inspect the exact content of logs to be uploaded at:
  <LOG_PATH>

  Press Enter to upload to Sentry, or Esc to cancel
